Fix DCC++ accessory address calculation for 'a' command.

dheap opened the pull request:
Addresses #6598 reported by @brookerg.

I don't have access to DCC++ hardware so would appreciate real-hardware testing. @brookerg has done some testing and reported success.

dheap commented:
I'm slightly surprised that @brookerg is the first to report this issue. This PR needs some real-world testing by others to ensure that there really is a problem with addressing in the current DCC++code/hardware/firmware combination.

Of course it's possible existing users have compensated by offsetting addresses and this change would have implications for them.

dheap commented:
https://travis-ci.org/JMRI/JMRI/jobs/498617634 timed out, restarted.

dheap commented:
I'm no C++ expert but I've downloaded the [DCC++ Base Station code](https://github.com/DccPlusPlus/BaseStation) and examined the format of the 'a' command and the code that converts it to raw DCC packets, comparing it with S9.2.1, our NmraPacket and NmraPacketTest code and my Packet Analyser sniffing of my NCE Power Pro 

I believe I've got it right, but it would be good to get confirmation from someone else with DCC++ hardware and typical accessory decoders  as well as @brookerg.

Pugwash1 commented:
@dheap Tried with original base station code and turnouts no longer need to be offset by 4.

Pugwash1 commented:
@dheap They are, however, still inverted.

pabender commented:
@pugwash inverted compared to what?

The thrown and closed definitions are really system specific, so if there isn’t anything to compare against, we just assign values to the two possible states.

Pugwash1 commented:
@pabender Sorry, youre correct, I knew that once,  We have to to do system prefix & turnout inverted search and replaces when swapping command stations, NCE & DCC are ON = Closed/normal and Digitrax  is ON = Thrown/reversed.

dheap commented:
DCC Packet 81 F8 79 (byte 2 bit 0 = 0) is produced by:
NCE Throttle ACCY 1 = R
JMRI NT1 Thrown
Reported as A0001R by NCE Packet Analyser in verbose mode
DCC++ 'a 1 0 0' command^, which is produced by:
JMRI 4.15.3 DT5 Closed
PR #6601 DT1 Closed
[DCC++ Wiki](https://github.com/DccPlusPlus/BaseStation/wiki/Commands-for-DCCpp-BaseStation) ACTIVATE: (0) (Deactivate, Off, Unthrown)


DCC Packet 81 F9 78 (byte 2 bit 0 = 1) is produced by:
NCE Throttle ACCY 1 = N
JMRI NT1 Closed
Reported as A0001N by NCE Packet Analyser in verbose mode
DCC++ 'a 1 0 1' command^, which is produced by:
JMRI 4.15.3 DT5 Thrown
PR #6601 DT1 Thrown
[DCC++ Wiki](https://github.com/DccPlusPlus/BaseStation/wiki/Commands-for-DCCpp-BaseStation) ACTIVATE: (1) (Activate, On, Thrown)


^ If my reading of lines 171-185 of [DCCpp_Uno/PacketRegister.cpp](https://github.com/DccPlusPlus/BaseStation/blob/master/DCCpp_Uno/PacketRegister.cpp) is correct, as I don't have DCC++ hardware.

So it looks like we are in line with each manufacturer's interpretation (they differ).

dheap commented:
I'm convinced enough to merge this, but we'll need a warning in the release notes.

dheap commented:
I'm working on eliminating an NPE when using Signal Heads with Traffic Monitor open.

dheap commented:
[AppVeyor](https://ci.appveyor.com/project/JMRI/jmri/builds/22736514) failed:

```
[00:27:58]     [junit] INFO  - No saved preferences, will open preferences window.  Searched for C:\Users\appveyor\AppData\Local\Temp\1\junit4265995322057629786\junit3668872628547332496\My_JMRI_Railroad.jmri\ProfileConfig.xml [Time-limited test] apps.Apps.<init>()
[00:27:58]     [junit] Exception in thread "TabbedPreferencesAction actionPerformed" java.lang.NullPointerException: Required nonnull default for apps.systemconsole.SystemConsolePreferencesManager does not exist.
[00:27:58]     [junit] 	at jmri.InstanceManager.getDefault(InstanceManager.java:203)
[00:27:58]     [junit] 	at apps.SystemConsoleConfigPanel.getPreferencesManager(SystemConsoleConfigPanel.java:240)
[00:27:58]     [junit] 	at apps.SystemConsoleConfigPanel.lambda$new$1(SystemConsoleConfigPanel.java:100)
[00:27:58]     [junit] 	at javax.swing.plaf.basic.BasicComboBoxUI.getDisplaySize(BasicComboBoxUI.java:1353)
[00:27:58]     [junit] 	at javax.swing.plaf.metal.MetalComboBoxUI.getMinimumSize(MetalComboBoxUI.java:331)
```

dheap commented:
This resolves #6598.

FrightRisk commented:
Hi. Is there documentation anywhere of how this is supposed to work. I've just started going through the commits to see what the JMRI changes were, but I don't see how DCC++ is working at all.  I thought the command was <a addr subaddr activate> And initially thought subaddr was the 2 part address scheme (addr 229 = 58 0 addr 47 = 12 2) But is that not the case? If this is the packet:

10AAAAAA 1AAACDDD

I don't see where DCC++ is ever setting C or the first two Ds.  It just flips the last bit on and off as if that is the activate/deactivate switch. But if you have this working, I must have missed something. What are all those test functions in the commit and how do you run them? I may be able to see what's happening with those.

dheap commented:
@FrightRisk
> Hi. Is there documentation anywhere of how this is supposed to work. I've just started going through the commits to see what the JMRI changes were, but I don't see how DCC++ is working at all. I thought the command was And initially thought subaddr was the 2 part address scheme (addr 229 = 58 0 addr 47 = 12 2) But is that not the case?

- Yes, based on my interpretation of the documentation of the 'a' command at the following places: [The DCCpp Wiki Commands](https://github.com/DccPlusPlus/BaseStation/wiki/Commands-for-DCCpp-BaseStation).
- [SerialCommand.cpp](https://github.com/DccPlusPlus/BaseStation/blob/master/DCCpp_Uno/SerialCommand.cpp).
- [PacketRegister.cpp](https://github.com/DccPlusPlus/BaseStation/blob/master/DCCpp_Uno/PacketRegister.cpp).

So accessory 2038 = 510 1. And similar... 

> If this is the packet:
> 
> 10AAAAAA 1AAACDDD
> 
> I don't see where DCC++ is ever setting C or the first two Ds. It just flips the last bit on and off as if that is the activate/deactivate switch. 

From [PacketRegister.cpp](https://github.com/DccPlusPlus/BaseStation/blob/master/DCCpp_Uno/PacketRegister.cpp):
```
void RegisterList::setAccessory(char *s) volatile{
  byte b[3];                      // save space for checksum byte
  int aAdd;                       // the accessory address (0-511 = 9 bits) 
  int aNum;                       // the accessory number within that address (0-3)
  int activate;                   // flag indicated whether accessory should be activated (1) or deactivated (0) following NMRA recommended convention
  
  if(sscanf(s,"%d %d %d",&aAdd,&aNum,&activate)!=3)
    return;
    
  b[0]=aAdd%64+128; // first byte is of the form 10AAAAAA, where AAAAAA represent 6 least signifcant bits of accessory address
```
- The `aAdd%64`loads the lower 6 bits into AAAAAA.
- The `+128` sets up the leading 10.
```
  b[1]=((((aAdd/64)%8)<<4) + (aNum%4<<1) + activate%2) ^ 0xF8;      // second byte is of the form 1AAACDDD, where C should be 1, and the least significant D represent activate/deactivate
```
- The `(((aAdd/64)%8)<<4)`loads the upper 6 bits into AAA.
- The `(aNum%4<<1)`loads the two most significant DD bits.
- The `activate%2`loads the least significant D bit.
- The overall `^ 0xF8` (an XOR) sets the leading 1, does the required ones complement of AAA and sets C=1.

Looks good to me when I compare it with JMRI's [NmraPacket.java ](https://github.com/JMRI/JMRI/blob/master/java/src/jmri/NmraPacket.java) and NMRA S9.2.1 Section D under **Basic Accessory Decoder Packet Format**.
> But if you have this working, I must have missed something.

I don't have DCC++ hardware but user reports at the time were in the affirmative.

> What are all those test functions in the commit and how do you run them? I may be able to see what's happening with those.

I'll address that in a separate comment later. Tea time here now...

dheap commented:
@FrightRisk 
> What are all those test functions in the commit and how do you run them? I may be able to see what's happening with those.

JMRI Tests run in a no-hardware mode and with no user-interactive  GUI. All those tests I added in this PR were to check that we are creating the correct 'a' commands. For Lights and Turnouts in JMRI we don't create any raw DCC packets, we just use the 'a', 't' or 'z' commands.

You can perform similar tests, with or without hardware(DCC++ Simulator mode) by issuing Turnout state change commands and having the DCC++ Traffic Monitor window open.

To run JMRI tests you need a local checkout of the JMRI code,  a JDK and an IDE or command line 'ant', as documented [here](https://www.jmri.org/help/en/html/doc/Technical/index.shtml).

This PR also addressed a JMRI problem with the parameters for the 'M' (send raw DCC packet) command we use for sending Signal Decoder Aspect packets and Signal Decoder OpsMode packets. The problem was that all raw packets created by JMRI's NmraPacket class include the 
error-detection byte. We didn't realise that DCC++ code appends its own error-detection byte. This was causing the DCC++ code to send a superfluous 0x00 byte (XOR of packet + checksum is always 0x00). But this has no relevance to the Accessory Decoder commands as they always use he 'a', 't' or 'z' commands.

FrightRisk commented:
Thank you! I've learned a lot more since my first post. I have been on the team updating DCC++. It is called DCC++ EX. But this 2 part addressing is a chore ;) My only remaining confusion is if DCC++ is properly interpreting the spec. Since Gregg hard coded C to be 1, it can never be 0. I thought the whole point of C was to "activate" and deactivate". Instead, he is using bit 0 of the second byte as activate. 

The spec says that you can have 4 output pairs and each pair can have two outputs. So in 1AAACSS0, C==1 activates it and C==0 deactivates what you are controlling. What is C controlling? I read it to be SS. So 0-3 output pairs. So that final 0 or 1 bit selects one of the two outputs in that pair. But in practice it seems that at least for turnouts, it makes sense that (249,0,0) is a "close" for and odd address (249,0,1) is "throw". I guess this happens because C is always one so it is turning ON two different things, which has the effect of turning one off and the other on. But with sounds, lights, etc. it confuses me.

Especially after reading 9.2.1, what makes sense to me would be a command that looked like this:

1AAACSSO (the second byte where A is address, C is control, SS is subaddress and O is output)

The DCC++ command would then be:

<A address sub_address port output activate>

<A address_1-511 subaddress_0-3 output_0-1 activate_0-1> where activate directly controls the C bit. Right now C is always 1 and bit zero( "O" for output), seems to be acting as activate/deactivate

IOW, I want the decoder at 223, the second of 4 output pairs, the first of its two outputs, and then activate it or deactivate it. I have direct control over 8 outputs with no complicated algebra or binary math.

I am pretty sure I am missing something by seeing it in the terms I've framed it with an 8 output controller instead of solenoid switches, but I can't seem to grasp it. Thanks for your help

bobjacobsen commented:
Broadly, here's a bit of back story.  When the NMRA started to write all this down, there as a _lot_ of corrupt internal NMRA politics to the process.  There are lots of places were a clear choice was subverted into an ambiguous mess.  In the specific case of accessory decoders, there were a number of manufacturers who wanted a specific message configuration that was analogous to C/MRI and other existing hardware.  Unfortunately, there was also another approach that a specific vendor desired.  Rather than make the decision that was in model railroading and model railroader's interest of a single well-documented and interoperable solution, there was a behind closed doors (literally) decision to write an ambiguous document that can be read to allow multiple forms.  And so now we have a mess.  

There are several common forms sent by command stations:

1) You set a turnout to +/-, normal/diverging, closed/thrown using the lower bit of the DDD address.  This comes in three sub-forms:
1A) Always have C set to 1 (there is no agreement whether 1 is on or off here, it's just a constant)
1B) Always have C set to 0
1C) Send the message first with 1, and then later with 0 (Digitrax throttles connect this to button down, button up for example)

This is the most common form  for commercial hardware in the US.  It's what JMRI normally does. (JMRI works through command stations in most cases, so the C bit often controlled by them; SPROG, Digitrax and a couple others have explicit control in JMRI)

1A and 1B can be confusing: There are turnout controllers where you send a turnout command to set the address; the controller just remembers it and looks for a repeat.  If you move one of those from a EasyDCC (type 1B) to a NCE (type 1A) system, they stop working.  Even 1C can be confusing: If you move the 1B EasyDCC controller to a Digitrax system, it still works, but it only moves when you _let go_ of the button.

There is another meaning, however:

2) DDD is an address that can have 8 values to address 8 sets of output terminals (though a particular hardware device may have fewer) and C sets the two directions of the turnout. This sounds more like modern electronics, and if you come from a computer background it seems obvious.  But it's not, and it wasn't, the traditional approach for model-railroad electronics where you want to drive current one way, drive current the other way and sometimes drive no current at all:  The 8 addresses solution when mapped directly to the output pins makes interfacing more complicated.  And back 20 years ago when every electrical element cost real money for a manufacturer, this solution ended up off to the side.

Devices that do this are still out there.  Some documentation talks about them _first_, which can cause confusion. Since this feels natural to programmers (even if the bits are laid out weirdly, due to another NMRA "negotiation"), people tend to latch on to this as how the standard must work.

OK, with that (really long) prolog, I recommend that JMRI and the DCC++ system together use the 1C approach (I have no opinion on the details of the messages, etc): Send two messages, separated a bit in time, with both C values.   You can see how JMRI does it for that command station protocol at[ line 150 of LnTurnout](https://github.com/JMRI/JMRI/blob/bd85548837050f1f46343ee01a0af65cceaefb38/java/src/jmri/jmrix/loconet/LnTurnout.java#L155). It sends the message to command C to 1/on, and then sets a timer to come back later for C to 0/off.  That feature is on by default, because it's the most reliable approach for new users, but it can be turned off if there's a lot of traffic for e.g. driving signal hardware that doesn't need it.

FrightRisk commented:
I love it! You gentlemen are patient educators and supporters. I appreciate the detail.

I'll have to look at how any front in addition to JMRI handle this so we don't break it in DCC++. One thought is to just change it and make sure the code in JMRI matches. Another is to add another command, <a> would just keep doing it the way it does now, which corresponds to your version 1A above, and a new command <A> would allow control over C for enable/disable.

The only remaining question is how to handle those 3 bits that control the output. In your final recommendation, you seemed to want to drop the "3 bits give me up to 8 outputs I can control with the C1 bit" and have it be the same as we have now, where the 0 bit is still closed/thrown for the next 2 bits (4 outputs), but C1 enables it and then after a time out or an explicit C0 command.

I think it confuses people too because they aren't controlling twin coil machines, but turning on LEDs, sounds or whatever. And you are right, a programmer sees those 3 bits and an on/off bit and just goes down that path. The whole "make the address 9 bits and take the most significant 3 bits the 4th, 5th and 6th bit  of the next byte and then make just those 3 in 1s complement form" is beyond obtuse to me. It was like it was purposely done so that only a programmer could figure it out.

FrightRisk commented:
One more point, the packets don't match the value. I am seeing that offset of 3. I am wondering if that is why JMRI was designed the way it was before this patch. Or I am missing something?

Example:
this formula...
int sw_address=1275;
sw1=(sw_address+3)/4;  
sw2=(sw_address-(sw1*4))+3;

gives 319,2

Use this formula to reverse check it:

N = (sw1 - 1) * 4 + sw2 + 1;

N=319

The DCC++ code is:

const char *s = "319 2 1";
sscanf(s,"%d %d %d",&aAdd,&aNum,&activate)
b[0]=aAdd%64+128;    // get the 6 addr bits and prepend 10 for 10AAAAAA
b[1]=((((aAdd/64)%8)<<4) + (aNum%4<<1) + activate%2) ^ 0xF8; // Build the 1AAACDD0 byte

That gives: 10111111 10111101

2s complement bits 4-7 and we get 100. Stick that in front of the 6 bit address:

100111111 

then append the 10 for the 2nd byte port address from bits 1 and 2 and we get:

10011111110 but this is 1278, not 1275

dheap commented:
@FrightRisk 
I think you're missing something. I don't follow your calculations.

If I use JMRI V4.19.6 in DCC++ Simulator Mode and using Actions->Turnout Control->Turnout 1275, I see the following in the DCC++ Traffic Monitor:
```
16:21:42.638: [TX: a 319 2 1]   Accessory Decoder Cmd: 
	Address: 319
	Subaddr: 2
	State: ON
```
Using my programmer's calculator:
- With `b[0]=aAdd%64+128` with `aAdd=319`, I get `10111111`.
- With `b[1]=((((aAdd/64)%8)<<4) + (aNum%4<<1) + activate%2) ^ 0xF8`, I get `10111101`.
That gives `10111111 10111101` or `BF BD`.

With My NCE Power Pro system and a connected Track DCC Packet Analyser, for the two states I see:

```
A1275N +++
BF BD 02 +++
A1275R +++
BF BC 03
```
Note that in NCE/JMRI terminology N=Closed, R=Thrown.  (DCC++ appears to take the opposite interpretation.)

I'm convinced the original JMRI and DCC++ code are correct.

I think your reverse calculation may be failing at the "2s complement". S9.2.1 says "ones complement".

FrightRisk commented:
Great! We both get the same numbers for the packet data. So I must be confused on my last calculation.

10111111 10111101

So how do you decide this? I meant 1s complement. Taking bits 5,6,7 of byte 2 (the AAA address bits) I see 011. The 1s complement is 100, right? So then don't you put 100 in front of the first 6 bits of byte 1?

100 111111 then add the 2 port bits (1 and 2 from byte 2)

100 111111 10 

10011111110 == 1278

So my mistake is here. What am I doing wrong?

dheap commented:
Because:
- Accessory 1 is `boardAddress=1` and `outputIndex=0`
- Accessory 2 is `boardAddress=1` and `outputIndex=1`
- Accessory 3 is `boardAddress=1` and `outputIndex=2`
- Accessory 4 is `boardAddress=1` and `outputIndex=3`
- Accessory 5 is `boardAddress=2` and `outputIndex=0`
- Accessory 6 is `boardAddress=2` and `outputIndex=1`
- Accessory 7 is `boardAddress=2` and `outputIndex=2`
- Accessory 8 is `boardAddress=2` and `outputIndex=3`

So you have to subtract 1 from `boardAddress` and add 1 to and `outputIndex`

> Why subtract 1 and then add 1 at the end?

- You are subtracting 1 from `boardAddress` then shifting `boardAddress` left 2.
- You are adding 1 to `outputIndex `.

dheap commented:
@FrightRisk 
Just to clarify, those Packet Grabs I posted above didn't involve `NmraPacket` at all. I just used JMRI to tell the NCE system to throw Accessory 1275 and it built the packets. In NCE-speak the transactions were:
```
11:25:08.745: [AD 04 FB 03 00]   Accessory 1275 Normal direction (ON)
11:25:08.746: [21]   Reply: 21
11:25:15.212: [AD 04 FB 04 00]   Accessory 1275 Reverse direction (OFF)
11:25:15.212: [21]   Reply: 21
```
I will do the packet grab again it using the NCE throttle keypad SELECT ACCY button on the NCE Pro Cab to throw the Accessory:
```
A1275N +++
BF BD 02 +++
A1275R +++
BF BC 03 +++
```
Just to emphasise that `NmraPacket` wasn't involved.

FrightRisk commented:
Thank you. That makes it all clear now :)

dheap commented:
@FrightRisk 
See earlier exchanges in this thread by @Pugwash1 and @pabender. There really isn't a consistent sense that "most systems" use.

Others can advise, but I'd be cautious about making a change as existing DCC++ users who move to your code will need to update every Turnout in their Table.

> So DCC++ may need an option to "invert output"

JMRI already has an `Inverted` checkbox in the Turnout Table so I don't think its necessary for you to duplicate that functionality in the DCC++ code.

I agree that my "incorrect address" calculation change was also disruptive but necessary as it was an error, not a matter of manufacturer-specific choice. I put a warning in the JMRI release notes to warn users of the need to change address.

dheap commented:
@FrightRisk 
A word of caution about edits. The settings for this group are such that edits do not trigger a resend of the comment to those subscribed by email**. So over half of the content of [this comment](https://github.com/JMRI/JMRI/pull/6601#issuecomment-629653714) has not been seen by those following by email. That means my reply that would be missing context.

** One of the groups I subscribe to (Loksound) does email edits. Increases the traffic a bit, but keeps everyone in the picture.

bobjacobsen commented:
> ** One of the groups I subscribe to (Loksound) does email edits. Increases the traffic a bit, but keeps everyone in the picture.

@dheap I looked through the settings and didn't see anything that would turn this on. Do know the name of the option or where it is?

dheap commented:
@bobjacobsen I just realised this is GitHub, not groups.io. Senior moment...
